rfcs/2021-11-22-design-tokens.md

```diff
@@ -0,0 +1,106 @@
+
+2. Once the SSOT is determined, tokens should be created in the appropriate format. At minimum, tokens should also be populated to form a set required to serve the current consuming applications. Ongoing work could then be ticketed to fill out the entire token library. These first tokens would represent the most basic items such as spacing, color, type attributes, breakpoints, animation easing values, border thickness, etc.. An example of a color token setup might resemble:
+
+   ```
+   tokens.git/core/colors.json
+   {
+   	"lightblue": {
+   		"100": "rgb(207, 239, 252)",
+   		"200": "rgb(113, 202, 243)",
+   		"300": "rgb(43, 199, 252)",
+   		"transparent": {
+   			"50": "rgba(43, 199, 252, 50)",
+   			"75": "rgba(43, 199, 252, 75)"
+   		}
+   	}
+   }
+   ```
+
+   > The aliases design tokens are mapped to in the various consuming systems need not match up directly one another initially. The token is a standard atomic unit which can be referenced in many systems using many aliases but should be applied as high up in the dependency tree as possible.
+
+3. When the first dataset is completed and ready for use, a tool stored in the NPM package would need to be created to convert these tokens into new formats, initially Sass. There are several existing tools which can do this such as [Style Dictionary](https://github.com/amzn/style-dictionary) (format is [easily utilized by the DSM](https://support.invisionapp.com/hc/en-us/articles/360049483032)), [Theo](https://github.com/salesforce-ux/theo), [JSON -> CSS](https://github.com/rlapoele/json-to-scss) and others.
+
+   I believe the best option would be to write our own [Webpack Loader](https://webpack.js.org/contribute/writing-a-loader/) to convert these tokens to top level Sass files.
+
+   1. In the case of the DS-Core, we would convert these tokens into variables which would be utilized by files in our /src/styles/settings/variables scss files and font.scss file initially. Since we would not be changing these files aside from referencing the token variable names in the tokens NPM package instead of providing hard coded values, the compiled output of these files should be compared to detect any breaking changes. One advantage to using a custom webpack loader in our case is that we could do things like generate the list of @font-face includes based on what tokens are available, slimming overall DS package size.
+   2. In the case of Sketch and the DSM, the tool chosen/created would be extended to output libraries for the DSM (if needed). Sketch and these libraries would also be versioned and live in the token package. These would be made available as release download libraries with which to work from. The [Sketch file format](https://developer.sketch.com/file-format/) is very easy to work with, as it is already made up of JSON files.
```

Comment:
> The [Sketch file format](https://developer.sketch.com/file-format/) is very easy to work with, as it is already made up of JSON files.

I'm curious to know more about this and how the high level tokens could map to color variables and layer styles in Sketch?

Reply:
That's a big ticket but basically it involves transforming the token JSON into JSON usable by Sketch and bundling it up in a zip file (the basic definition of what a Sketch file is) There are a lot of details there to work out before I can answer that. There are typescript and vanilla JS libraries to write the settings and there are even methods to export React components to Sketch, so that could be worked out over two sprints or so.

Reply:
How would you feel about leveraging a Sketch plugin like Puzzle tokens to pull tokens into sketch? https://github.com/ingrammicro/puzzle-tokens

Reply:
I don't like that it adds a software dependence where we could generate native Sketch files that would offer all that functionality and more with our own solution. I'm not against it, I just don't like the dependency.